packages/block-editor/src/components/media-replace-flow/index.js

```diff
@@ -0,0 +1,224 @@
+/**
+ * WordPress dependencies
+ */
+import { useState, createRef } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+import { speak } from '@wordpress/a11y';
+import {
+	FormFileUpload,
+	NavigableMenu,
+	MenuItem,
+	Toolbar,
+	Button,
+	Popover,
+	withNotices,
+} from '@wordpress/components';
+import {
+	LEFT,
+	RIGHT,
+	UP,
+	DOWN,
+	BACKSPACE,
+	ENTER,
+} from '@wordpress/keycodes';
+import { useSelect } from '@wordpress/data';
+import { compose } from '@wordpress/compose';
+
+/**
+ * Internal dependencies
+ */
+import MediaUpload from '../media-upload';
+import MediaUploadCheck from '../media-upload/check';
+import LinkEditor from '../url-popover/link-editor';
+import LinkViewer from '../url-popover/link-viewer';
+
+const MediaReplaceFlow = (
+	{
+		mediaURL,
+		allowedTypes,
+		onSelect,
+		onSelectURL,
+		noticeOperations,
+		name = __( 'Replace' ),
+		multiple = false,
+	}
+) => {
+	const [ showURLInput, setShowURLInput ] = useState( false );
+	const [ showEditURLInput, setShowEditURLInput ] = useState( false );
+	const [ mediaURLValue, setMediaURLValue ] = useState( mediaURL );
+	const [ showMediaReplaceOptions, setShowMediaReplaceOptions ] = useState( false );
+	const { mediaUpload } = useSelect( ( select ) => {
+		const { getSettings } = select( 'core/block-editor' );
+
+		return {
+			mediaUpload: getSettings().mediaUpload,
+		};
+	}, [] );
+	const editMediaButtonRef = createRef();
+
+	const stopPropagation = ( event ) => {
+		event.stopPropagation();
+	};
+
+	const stopPropagationRelevantKeys = ( event ) => {
+		if ( [ LEFT, DOWN, RIGHT, UP, BACKSPACE, ENTER ].indexOf( event.keyCode ) > -1 ) {
+			// Stop the key event from propagating up to ObserveTyping.startTypingInTextField.
+			event.stopPropagation();
+		}
+	};
+
+	const selectMedia = ( media ) => {
+		onSelect( media );
+		setMediaURLValue( media.url );
+		setShowMediaReplaceOptions( false );
+		speak( __( 'The media file has been replaced' ) );
+	};
+
+	const onUploadError = ( message ) => {
+		noticeOperations.removeAllNotices();
+		noticeOperations.createErrorNotice( message );
+	};
+
+	const uploadFiles = ( event ) => {
+		const files = event.target.files;
+		let setMedia;
+		if ( multiple ) {
+			setMedia = selectMedia;
+		} else {
+			setMedia = ( [ media ] ) => {
+				setShowMediaReplaceOptions( false );
+				selectMedia( media );
+			};
+		}
+		mediaUpload( {
+			allowedTypes,
+			filesList: files,
+			onFileChange: setMedia,
+			onUploadError,
+		} );
+	};
+
+	const fileUploadButton = (
+		<FormFileUpload
```

Comment:
This only seems to be used one, so it looks like there's no advantage of assigning it to a variable. Might as well inline it.